RFC0003-sparse-roadmap/README.md

```diff
@@ -0,0 +1,1352 @@
+<!--watch-latex-md
+
+This document is processed by watch_latex_md.py program, see
+
+  https://github.com/Quansight/pearu-sandbox/latex_in_markdown/
+
+You can edit this document as you wish. You can also edit the LaTeX
+data in img elements, but only the content of `latex-data`:
+
+  1. To automatically update the LaTeX rendering in img element, edit
+     the file while watch_latex_md.py is running.
+
+  2. Never change the beginning (`<img latex-data="...`) and the end
+     (`...alt="latex">`) parts of the LaTeX img elements as these are
+     used by the watch_latex_md.py script.
+
+  3. Changes to other parts of the LaTeX img elements will be
+     overwritten.
+
+Enjoy LaTeXing!
+
+watch-latex-md:no-force-rerender
+-->
+
+# Roadmap for PyTorch Sparse Tensors
+
+
+|            |                 |
+| ---------- | --------------- |
+| Authors    | Pearu Peterson  |
+| Status     | Draft           |
+| Type       | Process         |
+| Created    | 2020-04-23      |
+| Resolution | TBD             |
+
+## Abstract
+
+The aim of this document is to define the future of PyTorch sparse
+tensors. We propose that
+> :large_blue_circle:<!--:proposal:--> a sparse tensor is a drop-in
+> replacement of dense tensors for all tensor operations.
+
+## Motivation and Scope
+
+PyTorch implements the *Tensor* object as the basic structure of
+storing data in a multi-dimensional array structure that can be
+processed on various computational devices such as GPUs as well as
+multicore CPUs efficiently by taking advantage of parallelism that
+these devices support. PyTorch has advanged in developing
+computational tools for various ML and AI algorithms that can be
+applied to so-called dense or strided tensors. However, the
+developement of tools applicable to sparse tensors has not passed the
+experimental stage after two or more years of development, dispite the
+fact that a sparse tensor would be a natural data storage format for
+many ML&AI applications where the data characterizes some pairwise
+relations between study subjects, or when analyzing only a small set
+of features that dominate in a big system, for instance. Choosing a
+sparse tensor format to store the data may lead to more performant
+programs due to the efficient uses of memory as well as computational
+resources.
+
+The starting point of this proposal and the far-reaching goal is that
+the PyTorch tensor represents a multi-dimensional array whereas the
+used data storage format is an implementation detail that users can
+select depending on their data to (indirectly) control the choise of
+computational tools that would be most efficient for users'
+applications. Switching a storage format of data to another data
```

Comment:
If I understand correctly, it's a non-goal to support some method very specific to sparse tensors, such as preconditioned iterative methods? Because there might be more stuff specific to sparse tensor (which may not be even applicable to dense tensor)-- such how to partitioning the data, choosing the right preconditioner and Krylov subspace method, etc?